core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source,
+) : NSInputStream(NSData()) {
```

Comment:
`SourceNSInputStream` does not implement `scheduleInRunLoop`, so there will an error on attempt to execute code like this:
```
val stream = Buffer().also { it.writeString("blablablabla") }.asNSInputStream()
        stream.scheduleInRunLoop(NSRunLoop.currentRunLoop(), NSRunLoopCommonModes)
```

```
 2023-07-17 19:00:27.543 test.kexe[60305:7494596] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -scheduleInRunLoop:forMode: only defined for abstract class.  Define -[Test_kobjc0 scheduleInRunLoop:forMode:]!'
```

And some modern frameworks (like `NSURLSession`) still work with run loops. 

I'm not an ObjC/Swift expert, but given what I was told, inheriting form `NSInputStream`/`NSOutputStream` seems tricky.

Reply:
I did note this, but I haven't looked into implementing this `NSStream` API, as my use-case and others that just use `NSInputStream`'s `read()` API directly, don't require it. I can look into implementing, if it's not too much work.

Can you give an example of how this API is used in `NSURLSession`?

Reply:
Great, thanks for the test case. I'll look into implementing the delegate and run loop functionality. I ran across that blog post (not a lot written on the subject of `NSInputStream` subclasses), but didn't read it fully at the time. Interestingly, the author, BJ, is a local iOS dev I'm in a Slack group with. Maybe I'll reach out to him if I run into questions.

Reply:
I implemented the `NSStreamDelegate` and `scheduleInRunLoop()` APIs for both `SourceNSInputStream` and `SinkNSOutputStream`. Let me know if there's anything else I've missed.

Reply:
Thank you for fixing it, that's a terrific job!

Reply:
But it will take some time for me to get into details and review the change.